tests/TestTotp.cpp

```diff
@@ -83,3 +83,58 @@ void TestTotp::testTotpCode()
     output = QTotp::generateTotp(seed, time, 8, 30);
     QCOMPARE(output, QString("69279037"));
 }
+
+void TestTotp::testEncoderData()
+{
+    for (QMap<quint8, QTotp::Encoder>::const_iterator iter = QTotp::encoders.cbegin(); iter != QTotp::encoders.cend(); ++iter) {
+        const QTotp::Encoder& enc = iter.value();
+        QVERIFY2(enc.digits != 0,
+            qPrintable(QString("Custom encoders cannot have zero-value for digits field: %1(%2)").arg(enc.name).arg(iter.key())));
+        QVERIFY2(!enc.name.isEmpty(),
+            qPrintable(QString("Custom encoders must have a name: %1(%2)").arg(enc.name).arg(iter.key())));
+        QVERIFY2(!enc.shortName.isEmpty(),
+            qPrintable(QString("Custom encoders must have a shortName: %1(%2)").arg(enc.name).arg(iter.key())));
+        QVERIFY2(QTotp::shortNameToEncoder.contains(enc.shortName),
+            qPrintable(QString("No shortNameToEncoder entry found for custom encoder: %1(%2) %3").arg(enc.name).arg(iter.key()).arg(enc.shortName)));
+        QVERIFY2(QTotp::shortNameToEncoder[enc.shortName] == iter.key(),
+            qPrintable(QString("shortNameToEncoder doesn't reference this custome encoder: %1(%2) %3").arg(enc.name).arg(iter.key()).arg(enc.shortName)));
+        QVERIFY2(QTotp::nameToEncoder.contains(enc.name),
+            qPrintable(QString("No nameToEncoder entry found for custom encoder: %1(%2) %3").arg(enc.name).arg(iter.key()).arg(enc.shortName)));
+        QVERIFY2(QTotp::nameToEncoder[enc.name] == iter.key(),
+            qPrintable(QString("nameToEncoder doesn't reference this custome encoder: %1(%2) %3").arg(enc.name).arg(iter.key()).arg(enc.shortName)));
+    }
+
+    for (QMap<QString, quint8>::const_iterator iter = QTotp::nameToEncoder.cbegin(); iter != QTotp::nameToEncoder.cend(); ++iter) {
```

Comment:
Use range-based for on QTotp::nameToEncoder.keys().

Reply:
I'm happy to change it (especially since the data set is small).

IMO, that's one that Qt got wrong. Doing an allocation to create a temporary list of keys, then iterating over the keys then doing a map lookup to get the value makes what should be an O(n) operation into an O(n*log(n)) operation. But Qt doesn't provide a way to iterate over the key/value pair in O(n) except by using the iterators in a way that isn't compatible with range-based for loops.

Reply:
The keys() approach is a lot more readable, though. Not too happy with either solution. I guess it's fine as is.